drivers/flash/flash_sam.c

```diff
@@ -1,404 +1,492 @@
- */
+#define SAM_FLASH_WRITE_PAGE_SIZE (512)
 
+typedef void (*sam_flash_irq_init_fn_ptr)(void);
 
-/*
- * We only use block mode erases. The datasheet gives a maximum erase time
- * of 200ms for a 8KiB block.
- */
-#define SAM_FLASH_TIMEOUT_MS 220
-#if defined(IFLASH0_PAGE_SIZE)
-#define IFLASH_PAGE_SIZE IFLASH0_PAGE_SIZE
-#endif
-
-struct flash_sam_dev_cfg {
+struct sam_flash_config {
 	Efc *regs;
+	sam_flash_irq_init_fn_ptr irq_init;
+	off_t area_address;
+	off_t area_size;
+	struct flash_parameters parameters;
+	struct flash_pages_layout *pages_layouts;
+	size_t pages_layouts_size;
 };
 
-struct flash_sam_dev_data {
-	struct k_sem sem;
+struct sam_flash_erase_data {
+	off_t section_start;
+	size_t section_end;
+	bool succeeded;
 };
 
-static const struct flash_parameters flash_sam_parameters = {
-	.write_block_size = FLASH_WRITE_BLK_SZ,
-	.erase_value = 0xff,
+struct sam_flash_data {
+	const struct device *dev;
+	struct k_spinlock lock;
+	struct sam_flash_erase_data erase_data;
+	struct k_sem ready_sem;
 };
 
-static int flash_sam_write_protection(const struct device *dev, bool enable);
+static inline void sam_flash_mask_ready_interrupt(const struct sam_flash_config *config)
+{
+	Efc *regs = config->regs;
+
+	regs->EEFC_FMR &= ~EEFC_FMR_FRDY;
+}
 
-static inline void flash_sam_sem_take(const struct device *dev)
+static inline void sam_flash_unmask_ready_interrupt(const struct sam_flash_config *config)
 {
-	struct flash_sam_dev_data *data = dev->data;
+	Efc *regs = config->regs;
 
-	k_sem_take(&data->sem, K_FOREVER);
+	regs->EEFC_FMR |= EEFC_FMR_FRDY;
 }
 
-static inline void flash_sam_sem_give(const struct device *dev)
+static void sam_flash_isr(const struct device *dev)
 {
-	struct flash_sam_dev_data *data = dev->data;
+	struct sam_flash_data *data = dev->data;
+	const struct sam_flash_config *config = dev->config;
 
-	k_sem_give(&data->sem);
+	sam_flash_mask_ready_interrupt(config);
+	k_sem_give(&data->ready_sem);
 }
 
-/* Check that the offset is within the flash */
-static bool flash_sam_valid_range(const struct device *dev, off_t offset,
-				  size_t len)
+static int sam_flash_section_wait_until_ready(const struct device *dev)
 {
-	if (offset > CONFIG_FLASH_SIZE * 1024) {
-		return false;
+	struct sam_flash_data *data = dev->data;
+	const struct sam_flash_config *config = dev->config;
+	Efc *regs = config->regs;
+
+	k_sem_reset(&data->ready_sem);
+	sam_flash_unmask_ready_interrupt(config);
+
+	if (k_sem_take(&data->ready_sem, K_MSEC(500)) < 0) {
+		LOG_ERR("Command did not execute in time");
+		return -EFAULT;
 	}
 
-	if (len && ((offset + len - 1) > (CONFIG_FLASH_SIZE * 1024))) {
-		return false;
+	if (regs->EEFC_FSR & EEFC_FSR_FCMDE) {
+		LOG_ERR("Invalid command requested");
+		return -EPERM;
 	}
 
-	return true;
+	if (regs->EEFC_FSR & EEFC_FSR_FLOCKE) {
+		LOG_ERR("Tried to modify locked region");
+		return -EPERM;
+	}
+
+	if (regs->EEFC_FSR & EEFC_FSR_FLERR) {
+		LOG_ERR("Programming failed");
+		return -EPERM;
+	}
```

Comment:
Flags in `EEFC_FSR` are cleared when read.

```c
	uint32_t status = regs->EEFC_FSR;

	if (status & EEFC_FSR_FCMDE) {
		LOG_ERR("Invalid command requested");
		return -EPERM;
	}
	// ...
```

Reply:
@bjarki-trackunit this still needs to be fixed IMO

Reply:
Fixed with
https://github.com/zephyrproject-rtos/zephyr/blob/a06dfce570b5f7a56c75cf757212e9455bd375ad/drivers/flash/flash_sam.c#L100-L136